### PR TITLE
refactor: generalize some code of vis library

### PR DESCRIPTION
**Pull request by dvmoritzschoefl**
Generalises the vis library

### CODE REVIEW

**Review by dvzacharycutler — APPROVED** [review]
Looks great to me, these are very helpful changes. I disabled the sankey vis for now, as it still requires a few things like properly resizing/converting to mantine, but id still merge for the other very useful changes.

**Comment by dvmoritzschoefl** [comment]
> @dvzacharycutler I tested this PR in Ordino Enterprise and get the following compile error:
> 
> ```
> webpack compiled with 2 warnings
> ERROR in ./src/store/workbenchReducer.ts:79:51
> TS2339: Property 'numColumnsSelected' does not exist on type 'Partial<IVisConfig>'.
>   Property 'numColumnsSelected' does not exist on type 'Partial<ISankeyConfig>'.
>     77 |
>     78 |     if (state.workbenches[workbenchIndex].visConfig) {
>   > 79 |       state.workbenches[workbenchIndex].visConfig.numColumnsSelected = [];
>        |                                                   ^^^^^^^^^^^^^^^^^^
>     80 |       if (state.workbenches[workbenchIndex].visConfig.type === ESupportedPlotlyVis.BAR) {
>     81 |         (<IBarConfig>state.workbenches[workbenchIndex].visConfig).catColumnSelected = null;
>     82 |       } else if (state.workbenches[workbenchIndex].visConfig.type === ESupportedPlotlyVis.HEXBIN) {
> 
> Found 1 error in 16917 ms.
> ```
> 
> I used the package.json resolutions to install this branch:
> 
> ```js
>   "resolutions": {
>     "@types/react": "~18.2.0",
>     "@types/react-dom": "~18.2.0",
>     "react": "~18.2.0",
>     "react-dom": "~18.2.0",
>     "visyn_core": "git+ssh://git@github.com/datavisyn/visyn_core.git#mh/sankey"
>   },
> ```
> 
> Could you please have a look what needs to be changed in Ordino Enterprise and raise another PR there if necessary.

Yes I will take a look at it

**Comment by dvmoritzschoefl** [comment]
I cannot approve this PR but I reviewed the changes made by zach and they are ok

File: src/vis/provider/Provider.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  type: string,
+
+  /** The main vis renderer. Required in all visualizations. */
+  renderer: (props: ICommonVisProps<T>) => React.JSX.Element,
+
+  /** The sidebar renderer. Required in all visualizations. */
+  sidebarRenderer: (props: ICommonVisSideBarProps<T>) => React.JSX.Element,
+
+  mergeConfig: (columns: VisColumn[], config: T) => T,
+) {
```
Comment (thinkh) [inline src/vis/provider/Provider.tsx:17]:
```suggestion
): GeneralVis<T> {
```

File: src/vis/provider/Provider.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  renderer: (props: ICommonVisProps<T>) => React.JSX.Element;
+  sidebarRenderer: (props: ICommonVisSideBarProps<T>) => React.JSX.Element;
+  mergeConfig: (columns: VisColumn[], config: T) => T;
+}
+
+export const visMap: { [key: string]: GeneralVis<BaseConfig> } = {};
+
+export function getVisByConfig<T extends BaseConfig>(config: T) {
+  return visMap[config.type] as GeneralVis<T>;
+}
```
Comment (thinkh) [inline src/vis/provider/Provider.tsx:40]:
```suggestion
export function getVis<T extends BaseConfig = BaseConfig>(type: T["type"]): GeneralVis<T> | null {
  return visMap[type] as GeneralVis<T>;
}
```

alternatively, keep the original function.

File: src/vis/provider/utils.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { ViolinVisSidebar } from '../violin/ViolinVisSidebar';
+import { registerVis } from './Provider';
+
+export function registerAllVis() {
+  registerVis(ESupportedPlotlyVis.SCATTER, ScatterVis, ScatterVisSidebar, scatterMergeDefaultConfig);
+  registerVis(ESupportedPlotlyVis.BAR, BarVis, BarVisSidebar, barMergeDefaultConfig);
+  registerVis(ESupportedPlotlyVis.VIOLIN, ViolinVis, ViolinVisSidebar, violinMergeDefaultConfig);
+  registerVis(ESupportedPlotlyVis.HEXBIN, HexbinVis, HexbinVisSidebar, hexinbMergeDefaultConfig);
+  registerVis(ESupportedPlotlyVis.SANKEY, SankeyVis, SankeyVisSidebar, sankeyMergeDefaultConfig);
+}
```
Comment (thinkh) [inline src/vis/provider/utils.ts:23]: Move code to EagerVis in VisContainer and remove this utils.ts

File: src/vis/provider/index.tsx [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export * from './Provider';
+export * from './utils';
```
Comment (thinkh) [inline src/vis/provider/index.tsx:2]: Remove
```suggestion
```

File: src/vis/VisContainer.tsx
Comment (thinkh) [inline src/vis/VisContainer.tsx:1]: Rename to EagerVis.tsx

**Comment by dvmoritzschoefl** [comment]
Should be closed after https://github.com/datavisyn/visyn_core/pull/76 is merged

**Comment by thinkh** [comment]
Closed in favor of https://github.com/datavisyn/visyn_core/pull/76.

**Pull request closed without merging**